File: exports/src/exports/dossier.py
# SPDX-License-Identifier: Apache-2.0
# Copyright (C) 2026 The SIG project. Code is Apache-2.0; data and documentation
# carry per-artifact licences — see LICENSE and docs/2_canonical_design_spec.md §42.
"""The local dossier content contract (§39.2, SIG-UI-010..015) — the P06.1 slice
renderer.

This is a **minimal, slice-proof** renderer whose only job is to prove that J-1
renders end to end before the epistemic visual language exists. The production
dossier surface — the real owner of SIG-UI-010..015, with the full epistemic
language, a11y/no-JS, and editorial standards — is **P15.2, which supersedes this
renderer** (see the ticket scope note and ADR-032). Do not gold-plate it.

What it enforces from §39.2:

* **SIG-UI-010** the twelve sections, in the exact order.
* **SIG-UI-011** "what we don't know" is not an appendix: it appears in the
  summary, the print export, and the API (JSON) form.
* **SIG-UI-012** an explicit incompleteness banner naming the number of
  unresearched fields and stating absence-is-not-evidence-of-absence.
* **SIG-UI-013** a print/PDF path: paginated, with sources, and with the as-of
  date and permalink on every page.
* **SIG-UI-014** every material figure is expandable to its reconciliation: the
  rule that fired, the competing claims, each source's tier and date, and a link
  to the document at its locator.
* **SIG-UI-015** `unknown` values are rendered as "unknown", never omitted.
"""

from __future__ import annotations

import html
import json
from dataclasses import dataclass
from datetime import date

# --- SIG-UI-010: the twelve sections, in order -------------------------------

SECTION_ORDER: tuple[tuple[str, str], ...] = (
    ("at_a_glance", "At a glance"),
    ("what_is_deployed", "What is deployed"),
    ("cost_and_expiry", "Cost and expiry"),
    ("who_else_can_see", "Who else can see the data"),
    ("configuration_and_retention", "Configuration and retention"),
    ("usage", "Usage"),
    ("where_the_hardware_is", "Where the hardware is"),
    ("policy", "Policy"),
    ("accountability_events", "Accountability events"),
    ("timeline", "Timeline"),
    ("what_we_dont_know", "What we don't know"),
    ("how_we_know_this", "How we know this"),
)
SECTION_IDS: tuple[str, ...] = tuple(sid for sid, _ in SECTION_ORDER)
SECTION_TITLES: dict[str, str] = dict(SECTION_ORDER)

# Statuses a field may carry when it has no resolved value (SIG-UI-015, §9.5).
NOT_RESEARCHED = "not_researched"
SEARCHED_NOT_FOUND = "searched_not_found"


@dataclass(frozen=True)
class DocumentRef:
    """A link to a document at its exact locator (SIG-UI-014, §D.4)."""

    source_family: str
    stable_locator: str
    locator: dict[str, object]
    capture_digest: str = ""
    excerpt: str = ""

    def as_json(self) -> dict[str, object]:
        return {
            "source_family": self.source_family,
            "stable_locator": self.stable_locator,
            "locator": self.locator,
            "capture_digest": self.capture_digest,
            "excerpt": self.excerpt,
        }


@dataclass(frozen=True)
class ReconClaim:
    """One competing claim behind a figure, with its tier and date (SIG-UI-014)."""

    value: object
    source_family: str
    reliability: str  # R tier
    weight: int | None  # W class
    observed_at: date
    document: DocumentRef

    def as_json(self) -> dict[str, object]:
        return {
            "value": self.value,
            "source_family": self.source_family,
            "reliability": self.reliability,
            "weight": self.weight,
            "observed_at": self.observed_at.isoformat(),
            "document": self.document.as_json(),
        }


@dataclass(frozen=True)
class Reconciliation:
    """The expandable reconciliation behind a material figure (SIG-UI-014)."""

    rule: str  # the rule that fired, e.g. "authoritative_source_wins -> W4"
    winning: ReconClaim
    competing: tuple[ReconClaim, ...] = ()
    note: str = ""

    def as_json(self) -> dict[str, object]:
        return {
            "rule": self.rule,
            "winning": self.winning.as_json(),
            "competing": [c.as_json() for c in self.competing],
            "note": self.note,
        }


@dataclass(frozen=True)
class Figure:
    """A material figure — always expandable to its reconciliation (SIG-UI-014)."""

    key: str
    label: str
    value: object
    reconciliation: Reconciliation
    unit: str = ""
    lower_bound: bool = False

    def as_json(self) -> dict[str, object]:
        return {
            "key": self.key,
            "label": self.label,
            "value": self.value,
            "unit": self.unit,
            "lower_bound": self.lower_bound,
            "reconciliation": self.reconciliation.as_json(),
        }


@dataclass(frozen=True)
class Row:
    """A non-figure fact. ``value=None`` renders as "unknown" (SIG-UI-015)."""

    label: str
    value: object | None
    status: str = "known"  # known | not_researched | searched_not_found
    document: DocumentRef | None = None
    note: str = ""

    def display_value(self) -> str:
        if self.value is not None:
            return str(self.value)
        return "unknown"

    def as_json(self) -> dict[str, object]:
        return {
            "label": self.label,
            "value": self.value,
            "display_value": self.display_value(),
            "status": self.status,
            "document": self.document.as_json() if self.document else None,
            "note": self.note,
        }


@dataclass(frozen=True)
class Section:
    section_id: str
    figures: tuple[Figure, ...] = ()
    rows: tuple[Row, ...] = ()

    @property
    def title(self) -> str:
        return SECTION_TITLES[self.section_id]

    def as_json(self) -> dict[str, object]:
        return {
            "id": self.section_id,
            "title": self.title,
            "figures": [f.as_json() for f in self.figures],
            "rows": [r.as_json() for r in self.rows],
        }


@dataclass(frozen=True)
class Gap:
    """One entry of "what we don't know" (SIG-UI-011)."""

    label: str
    status: str  # not_researched | searched_not_found
    note: str = ""

    def as_json(self) -> dict[str, object]:
        return {"label": self.label, "status": self.status, "note": self.note}


@dataclass(frozen=True)
class Dossier:
    subject_label: str
    jurisdiction: str
    as_of: date
    permalink: str
    sections: tuple[Section, ...]
    gaps: tuple[Gap, ...]
    unresearched_field_count: int
    source_families: tuple[str, ...] = ()

    def incompleteness_banner(self) -> str:
        """The explicit incompleteness banner (SIG-UI-012)."""
        return (
            f"This dossier has {self.unresearched_field_count} unresearched field(s). "
            "The absence of a row is not evidence of absence."
        )

    def validate(self) -> None:
        """Enforce the §39.2 section contract (SIG-UI-010/011)."""
        got = tuple(s.section_id for s in self.sections)
        if got != SECTION_IDS:
            raise ValueError(
                f"dossier sections must be exactly the §39.2 order (SIG-UI-010); got {got!r}"
            )


# --- SIG-UI-011: "what we don't know" in summary + API + print ---------------


def render_json(dossier: Dossier) -> dict[str, object]:
    """The API form. Carries the incompleteness banner and the gap list at the
    top level (the summary) AND inside the sections (SIG-UI-011)."""
    dossier.validate()
    return {
        "subject": dossier.subject_label,
        "jurisdiction": dossier.jurisdiction,
        "as_of": dossier.as_of.isoformat(),
        "permalink": dossier.permalink,
        "incompleteness_banner": dossier.incompleteness_banner(),
        "unresearched_field_count": dossier.unresearched_field_count,
        # "what we don't know" is a headline feature, in the summary AND its section.
        "what_we_dont_know": [g.as_json() for g in dossier.gaps],
        "sections": [s.as_json() for s in dossier.sections],
        "source_families": list(dossier.source_families),
    }


def render_json_str(dossier: Dossier) -> str:
    return json.dumps(render_json(dossier), indent=2, sort_keys=False)


# --- SIG-UI-013: the print/PDF path ------------------------------------------


def _e(text: object) -> str:
    return html.escape(str(text))


def _figure_html(fig: Figure) -> str:
    r = fig.reconciliation
    lb = " (lower bound)" if fig.lower_bound else ""
    rows = [
        "<details class='reconciliation'>",
        f"<summary>{_e(fig.label)}: <strong>{_e(fig.value)}</strong>"
        f"{_e(fig.unit and ' ' + fig.unit)}{lb}</summary>",
        f"<p class='rule'>Rule: {_e(r.rule)}</p>",
        f"<p class='note'>{_e(r.note)}</p>",
        "<table class='competing'><thead><tr>"
        "<th>value</th><th>source</th><th>tier</th><th>W</th><th>date</th><th>document</th>"
        "</tr></thead><tbody>",
    ]
    for c in (r.winning, *r.competing):
        rows.append(
            "<tr>"
            f"<td>{_e(c.value)}</td><td>{_e(c.source_family)}</td>"
            f"<td>{_e(c.reliability)}</td><td>{_e(c.weight)}</td>"
            f"<td>{_e(c.observed_at.isoformat())}</td>"
            f"<td><a href='{_e(c.document.stable_locator)}'>"
            f"{_e(c.document.locator)}</a></td>"
            "</tr>"
        )
    rows.append("</tbody></table></details>")
    return "\n".join(rows)


def _row_html(row: Row) -> str:
    doc = ""
    if row.document is not None:
        doc = f" <a class='src' href='{_e(row.document.stable_locator)}'>[source]</a>"
    cls = "" if row.value is not None else " class='unknown'"
    return (
        f"<li{cls}><span class='label'>{_e(row.label)}:</span> {_e(row.display_value())}{doc}</li>"
    )


def _section_html(section: Section) -> str:
    parts = [f"<h2>{_e(section.title)}</h2>"]
    for fig in section.figures:
        parts.append(_figure_html(fig))
    if section.rows:
        parts.append("<ul class='rows'>")
        parts.extend(_row_html(r) for r in section.rows)
        parts.append("</ul>")
    return "\n".join(parts)


def _page_footer_html(dossier: Dossier) -> str:
    # Present on EVERY page (SIG-UI-013): as-of date + permalink.
    return (
        "<footer class='page-footer'>"
        f"<span class='as-of'>As of {_e(dossier.as_of.isoformat())}</span>"
        f" · <a class='permalink' href='{_e(dossier.permalink)}'>{_e(dossier.permalink)}</a>"
        "</footer>"
    )


_PRINT_CSS = """
@page { size: letter; margin: 2cm; }
body { font-family: Georgia, 'Times New Roman', serif; color: #111; }
.page { page-break-after: always; }
.page:last-child { page-break-after: auto; }
.page-footer { border-top: 1px solid #999; margin-top: 1.5rem; padding-top: .4rem;
  font-size: .75rem; color: #444; }
.incompleteness { border: 2px solid #b00; padding: .6rem; margin: .8rem 0; font-weight: bold; }
.unknown { color: #666; font-style: italic; }
details.reconciliation { margin: .4rem 0; }
table.competing { border-collapse: collapse; font-size: .8rem; }
table.competing th, table.competing td { border: 1px solid #ccc; padding: .2rem .4rem; }
"""


def render_print_html(dossier: Dossier, *, sections_per_page: int = 3) -> str:
    """The print/PDF path (SIG-UI-013): paginated, with sources, and with the
    as-of date + permalink on every page. Print-to-PDF from a browser produces a
    council-ready document; a server-side PDF renderer is deferred to P15.2."""
    dossier.validate()
    footer = _page_footer_html(dossier)
    banner = f"<div class='incompleteness'>{_e(dossier.incompleteness_banner())}</div>"

    pages: list[str] = []
    # Page 1 header carries the title, the incompleteness banner, and a summary of
    # what we don't know (SIG-UI-011/012).
    gap_items = "\n".join(f"<li>{_e(g.label)} — <em>{_e(g.status)}</em></li>" for g in dossier.gaps)
    header = (
        f"<h1>{_e(dossier.subject_label)}</h1>"
        f"<p class='jurisdiction'>{_e(dossier.jurisdiction)}</p>"
        f"{banner}"
        "<section class='summary-gaps'><h2>What we don't know (summary)</h2>"
        f"<ul>{gap_items}</ul></section>"
    )

    chunks = [
        dossier.sections[i : i + sections_per_page]
        for i in range(0, len(dossier.sections), sections_per_page)
    ]
    for idx, chunk in enumerate(chunks):
        body = "\n".join(_section_html(s) for s in chunk)
        head = header if idx == 0 else ""
        pages.append(f"<div class='page'>\n{head}\n{body}\n{footer}\n</div>")

    return (
        "<!DOCTYPE html>\n<html lang='en'>\n<head>\n<meta charset='utf-8'>\n"
        f"<title>{_e(dossier.subject_label)} — SIG dossier</title>\n"
        f"<style>{_PRINT_CSS}</style>\n</head>\n<body>\n"
        + "\n".join(pages)
        + "\n</body>\n</html>\n"
    )


def page_count(
    html_doc: str, *, sections_per_page: int = 3, section_total: int = len(SECTION_IDS)
) -> int:
    """The number of print pages (for tests): ceil(sections / per_page)."""
    return -(-section_total // sections_per_page)


__all__ = [
    "SECTION_IDS",
    "SECTION_ORDER",
    "SECTION_TITLES",
    "NOT_RESEARCHED",
    "SEARCHED_NOT_FOUND",
    "DocumentRef",
    "ReconClaim",
    "Reconciliation",
    "Figure",
    "Row",
    "Section",
    "Gap",
    "Dossier",
    "render_json",
    "render_json_str",
    "render_print_html",
    "page_count",
]

File: web/src/lib/dossier.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright (C) 2026 The SIG project. Code is Apache-2.0; data and documentation
// carry per-artifact licences — see LICENSE and docs/2_canonical_design_spec.md §42.

/**
 * The local dossier — the project's primary public artifact (§39.2), as pure data
 * + logic. This is the PRODUCTION dossier content contract (SIG-UI-010..015); it
 * supersedes the P06.1 slice renderer (`exports/src/exports/dossier.py`, ADR-032),
 * rebuilt on P15.1's epistemic visual language and a11y/no-JS baseline.
 *
 * This module owns, colour-free (colour lives only in `styles/epistemic.css`):
 *   - the twelve §39.2 sections in their exact order + a validator (SIG-UI-010);
 *   - the "what we don't know" gap model, surfaced in the summary, the print export,
 *     and the API (SIG-UI-011);
 *   - the incompleteness banner (count of unresearched fields + the absence rule,
 *     SIG-UI-012);
 *   - the expandable reconciliation behind every material figure (SIG-UI-014);
 *   - the three action blocks the outline omits — `authorization`,
 *     `termination_mechanics`, `legal_regime` (SIG-UI-014a);
 *   - the derived `next_decision_date` (SIG-UI-014b), a STABLE wire name the renewal
 *     watch (P15.4, §39.5) keys its alerts on — treat it as an interface contract;
 *   - the Appendix-B content contract, where an `unknown` value renders as "unknown"
 *     rather than being omitted (SIG-UI-015).
 *
 * The rendering (`pages/dossier/*`) and the API endpoint (`dossier/[slug].json.ts`)
 * both derive from `renderDossierJson`, so the HTML summary, the print export, and
 * the JSON API can never drift out of the SIG-UI-011 contract.
 */

import { ABSENCE_KIND_META } from "./epistemic";
import type { AbsenceKind, CompetingClaim, Support } from "./epistemic";
import { beliefPinnedPermalink } from "./citation";
import type { AsOfEcho } from "./fixtures";
import type { AbsenceTaskParams } from "./task";

// --- SIG-UI-010: the twelve sections, in the exact §39.2 order ---------------

/** The twelve dossier sections, `[id, title]`, in the exact §39.2 order. */
export const SECTION_ORDER: readonly (readonly [string, string])[] = [
  ["at_a_glance", "At a glance"],
  ["what_is_deployed", "What is deployed"],
  ["cost_and_expiry", "Cost and expiry"],
  ["who_else_can_see", "Who else can see the data"],
  ["configuration_and_retention", "Configuration and retention"],
  ["usage", "Usage"],
  ["where_the_hardware_is", "Where the hardware is"],
  ["policy", "Policy"],
  ["accountability_events", "Accountability events"],
  ["timeline", "Timeline"],
  ["what_we_dont_know", "What we don't know"],
  ["how_we_know_this", "How we know this"],
] as const;

export const SECTION_IDS: readonly string[] = SECTION_ORDER.map(([id]) => id);
export const SECTION_TITLES: Record<string, string> = Object.fromEntries(SECTION_ORDER);

// --- Reconciliation behind a material figure (SIG-UI-014) --------------------

/**
 * The expandable reconciliation for a material figure: the rule that fired, the
 * competing claims (each with its source, tier, date, and a link to the document
 * at its supporting locator), and which claim won. Reuses the shared
 * `CompetingClaim` shape so the dossier's reconciliation and the contradiction
 * range render the same evidence (SIG-UI-009/014).
 */
export interface Reconciliation {
  /** The resolver rule that fired, e.g. "HIGHEST_TIER_WINS". */
  rule: string;
  /** The `claimId` of the winning claim within `claims`. */
  winningClaimId: string;
  claims: CompetingClaim[];
  /** A plain-language note (the local advocate is the design center). */
  note?: string;
}

/** A material figure — always expandable to its reconciliation (SIG-UI-014). */
export interface Figure {
  key: string;
  label: string;
  value: string | number;
  unit?: string;
  /** True when the value is a lower bound (e.g. mapped-device count, §D.2). */
  lowerBound?: boolean;
  support: Support;
  /** Independent evidence-class count behind the winning value (SIG-UI-003). */
  evidenceCount: number;
  /** Whether the value is contested (drives the persistent marker, SIG-UI-008). */
  contested: boolean;
  reconciliation: Reconciliation;
}

// --- The Appendix-B content contract row (SIG-UI-015) ------------------------

/**
 * A non-figure fact. `value === null` with no `absence` renders explicitly as
 * "unknown" — never omitted (SIG-UI-015). A row with an `absence` kind is a
 * clickable gap (rendered as the single hatch, SIG-UI-007) and, when
 * `NOT_RESEARCHED`, counts toward the incompleteness banner (SIG-UI-012).
 */
export interface Row {
  label: string;
  value: string | number | null;
  /** When set, this field is a gap: rendered as the absence hatch, clickable to a task. */
  absence?: AbsenceKind;
  /** For a taskable absence row, the subject/predicate the hatch links with. */
  subject_id?: string;
  predicate_id?: string;
  /** A link to the supporting document at its locator (SIG-UI-014). */
  documentUrl?: string;
  note?: string;
}

/** The explicit display string for a row value (SIG-UI-015: "unknown", not omitted). */
export function rowDisplayValue(row: Row): string {
  if (row.absence) return ABSENCE_KIND_META[row.absence].label;
  return row.value === null ? "unknown" : String(row.value);
}

export interface Section {
  section_id: string;
  figures?: Figure[];
  rows?: Row[];
}

/** One entry of "what we don't know" (SIG-UI-011); its kind is one of the four (§9.5). */
export interface Gap {
  label: string;
  kind: AbsenceKind;
  subject_id: string;
  predicate_id: string;
  /** For NO_EVIDENCE_FOUND: the named sources searched (§9.5, SIG-TIME-011). */
  sources_searched?: string[];
  note?: string;
}

// --- The three action blocks the outline omits (SIG-UI-014a) -----------------

/**
 * Who authorized the deployment, and how. "Approved 7–0 after public comment" and
 * "passed unopposed on the consent agenda" are politically opposite facts, so the
 * consent-agenda flag and the public-comment flag are first-class (SIG-UI-014a).
 */
export interface Authorization {
  approving_body: string | null;
  vote: string | null;
  consent_agenda: boolean | null;
  public_comment: boolean | null;
  date: string | null;
}

/**
 * The raw termination inputs. `next_decision_date` is DERIVED from these
 * (`resolveTermination`), never stored, so it can never disagree with the inputs.
 */
export interface TerminationInput {
  auto_renews: boolean;
  notice_window_days: number | null;
  expiry_date: string | null;
}

/** Termination mechanics with the derived decision date (SIG-UI-014a/b). */
export interface TerminationMechanics extends TerminationInput {
  /** The decision date, not the expiry date — surfaced wherever expiry is (SIG-UI-014b). */
  next_decision_date: string | null;
}

export interface LegalRegime {
  state_statute: string | null;
  local_ordinance: string | null;
  disclosure_duties: string[];
}

/**
 * The renewal decision date (SIG-UI-014b). An expiry date is the WRONG figure to
 * surface: a contract expiring 2027-04-02 with auto-renewal and a 90-day notice
 * window has a real deadline of 2027-01-02 — after which the decision is made by
 * default. So when the contract auto-renews, the decision date is the expiry minus
 * the notice window; otherwise the decision must be taken by the expiry itself.
 * The renewal watch (P15.4, §39.5) keys its alerts on this exact value.
 */
export function nextDecisionDate(t: TerminationInput): string | null {
  if (!t.expiry_date) return null;
  if (t.auto_renews && t.notice_window_days !== null) {
    return subtractDays(t.expiry_date, t.notice_window_days);
  }
  return t.expiry_date;
}

/** Resolve the raw termination inputs into the full block with its derived date. */
export function resolveTermination(t: TerminationInput): TerminationMechanics {
  return { ...t, next_decision_date: nextDecisionDate(t) };
}

/** Subtract whole days from an ISO date (UTC), returning an ISO `YYYY-MM-DD`. */
function subtractDays(isoDate: string, days: number): string {
  const ms = Date.parse(`${isoDate}T00:00:00Z`);
  if (Number.isNaN(ms)) throw new Error(`invalid ISO date: "${isoDate}"`);
  const d = new Date(ms - days * 86_400_000);
  return d.toISOString().slice(0, 10);
}

// --- The dossier itself ------------------------------------------------------

export interface Dossier {
  slug: string;
  subject_label: string;
  jurisdiction: string;
  asOf: AsOfEcho;
  rulesetVersion: string;
  sections: Section[];
  gaps: Gap[];
  source_families: string[];
  authorization: Authorization;
  termination: TerminationInput;
  legal_regime: LegalRegime;
}

/** The canonical page path for a dossier (trailing slash — SIG-UI-035). */
export function dossierPath(slug: string): string {
  return `/dossier/${slug}/`;
}

/** The canonical path of a dossier's print export. */
export function dossierPrintPath(slug: string): string {
  return `/dossier/${slug}/print/`;
}

/** The canonical path of a dossier's static JSON (API-form) endpoint. */
export function dossierJsonPath(slug: string): string {
  return `/dossier/${slug}.json`;
}

/**
 * The count of DISTINCT unresearched fields the incompleteness banner names
 * (SIG-UI-012): every `NOT_RESEARCHED` gap plus every section row that is a
 * `NOT_RESEARCHED` absence, deduplicated by `(subject_id, predicate_id)` so a field
 * elevated to the "what we don't know" headline AND shown in its section is counted
 * once. "No evidence found" is NOT unresearched — SIG looked — so it is excluded.
 */
export function unresearchedFieldCount(dossier: Dossier): number {
  const keys = new Set<string>();
  for (const g of dossier.gaps) {
    if (g.kind === "NOT_RESEARCHED") keys.add(`${g.subject_id}\u0000${g.predicate_id}`);
  }
  for (const s of dossier.sections) {
    for (const r of s.rows ?? []) {
      if (r.absence === "NOT_RESEARCHED") {
        keys.add(`${r.subject_id ?? dossier.slug}\u0000${r.predicate_id ?? r.label}`);
      }
    }
  }
  return keys.size;
}

/** The explicit incompleteness banner (SIG-UI-012): count + the absence rule. */
export function incompletenessBanner(dossier: Dossier): string {
  const n = unresearchedFieldCount(dossier);
  return (
    `This dossier has ${n} unresearched field${n === 1 ? "" : "s"}. ` +
    "The absence of a row is not evidence of absence."
  );
}

/** The belief-pinned permalink for a dossier page (SIG-UI-035). */
export function dossierPermalink(dossier: Dossier, origin?: string): string {
  return beliefPinnedPermalink({
    path: dossierPath(dossier.slug),
    title: dossier.subject_label,
    asOf: dossier.asOf,
    rulesetVersion: dossier.rulesetVersion,
    ...(origin ? { origin } : {}),
  });
}

/**
 * Validate the §39.2 section contract (SIG-UI-010): the sections MUST be exactly
 * the twelve ids in the canonical order. Throws otherwise, so a mis-ordered or
 * incomplete dossier can never render.
 */
export function validateDossier(dossier: Dossier): void {
  const got = dossier.sections.map((s) => s.section_id);
  const want = SECTION_IDS;
  const ok = got.length === want.length && got.every((id, i) => id === want[i]);
  if (!ok) {
    throw new Error(
      `dossier sections must be exactly the §39.2 order (SIG-UI-010); got ${JSON.stringify(got)}`,
    );
  }
}

// --- The API form (SIG-UI-011): one source of truth for all three surfaces ---

/**
 * The dossier's API (JSON) representation. Because the shell is static-first
 * (SIG-UI-036) it is emitted as a committed static endpoint at build time
 * (`dossier/[slug].json.ts`), not read from a live API — but the shape is the
 * `/v1` dossier contract. Critically, "what we don't know" appears BOTH at the
 * summary top level (`what_we_dont_know`) AND inside the sections, so the API,
 * the print export, and the HTML summary all satisfy SIG-UI-011.
 */
export function renderDossierJson(dossier: Dossier, origin?: string): Record<string, unknown> {
  validateDossier(dossier);
  const permalink = dossierPermalink(dossier, origin);
  return {
    subject: dossier.subject_label,
    jurisdiction: dossier.jurisdiction,
    as_of_world: dossier.asOf.as_of_world,
    as_of_belief: dossier.asOf.as_of_belief,
    ruleset_version: dossier.rulesetVersion,
    permalink,
    incompleteness_banner: incompletenessBanner(dossier),
    unresearched_field_count: unresearchedFieldCount(dossier),
    // "What we don't know" is a headline feature, at the summary top level AND
    // rendered inside its section on the page/print (SIG-UI-011).
    what_we_dont_know: dossier.gaps.map((g) => ({
      label: g.label,
      kind: g.kind,
      subject_id: g.subject_id,
      predicate_id: g.predicate_id,
      ...(g.sources_searched ? { sources_searched: g.sources_searched } : {}),
      ...(g.note ? { note: g.note } : {}),
    })),
    // The three action blocks the outline omits (SIG-UI-014a), with the derived
    // decision date (SIG-UI-014b) computed once, here.
    authorization: dossier.authorization,
    termination_mechanics: resolveTermination(dossier.termination),
    legal_regime: dossier.legal_regime,
    sections: dossier.sections.map((s) => ({
      id: s.section_id,
      title: SECTION_TITLES[s.section_id],
      figures: (s.figures ?? []).map(figureJson),
      rows: (s.rows ?? []).map(rowJson),
    })),
    source_families: dossier.source_families,
  };
}

function figureJson(fig: Figure): Record<string, unknown> {
  const winning = fig.reconciliation.claims.find((c) => c.claimId === fig.reconciliation.winningClaimId);
  return {
    key: fig.key,
    label: fig.label,
    value: fig.value,
    unit: fig.unit ?? "",
    lower_bound: fig.lowerBound ?? false,
    support: fig.support,
    evidence_count: fig.evidenceCount,
    contested: fig.contested,
    reconciliation: {
      rule: fig.reconciliation.rule,
      note: fig.reconciliation.note ?? "",
      winning_claim_id: fig.reconciliation.winningClaimId,
      // The winning claim first, then the competing claims — each with tier, date,
      // and a document link at its locator (SIG-UI-014).
      claims: fig.reconciliation.claims.map((c) => ({
        claim_id: c.claimId,
        value: c.value,
        source: c.source,
        tier: c.tier,
        date: c.date,
        document_url: c.documentUrl,
        winning: c.claimId === fig.reconciliation.winningClaimId,
        ...(c.differentQuantityNote ? { different_quantity_note: c.differentQuantityNote } : {}),
      })),
      winning_present: winning !== undefined,
    },
  };
}

function rowJson(row: Row): Record<string, unknown> {
  return {
    label: row.label,
    value: row.value,
    display_value: rowDisplayValue(row),
    ...(row.absence ? { absence_kind: row.absence } : {}),
    ...(row.documentUrl ? { document_url: row.documentUrl } : {}),
    ...(row.note ? { note: row.note } : {}),
  };
}

/**
 * Every taskable absence a dossier links a hatch to — its "what we don't know"
 * gaps plus every section row that is an absence — as `AbsenceTaskParams`. The
 * task-intake route (`pages/task/new/[slug].astro`) unions these into its
 * `getStaticPaths` so every clickable dossier gap resolves to a real, pre-generated
 * intake page with no client JavaScript (SIG-UI-007, SIG-UI-036/037).
 */
export function dossierTaskableAbsences(dossier: Dossier): AbsenceTaskParams[] {
  const fromGaps: AbsenceTaskParams[] = dossier.gaps.map((g) => ({
    subject_id: g.subject_id,
    predicate_id: g.predicate_id,
    absence_kind: g.kind,
    predicate_label: g.label,
  }));
  const fromRows: AbsenceTaskParams[] = dossier.sections.flatMap((s) =>
    (s.rows ?? [])
      .filter((r): r is Row & { absence: AbsenceKind } => r.absence !== undefined)
      .map((r) => ({
        subject_id: r.subject_id ?? dossier.slug,
        predicate_id: r.predicate_id ?? r.label,
        absence_kind: r.absence,
        predicate_label: r.label,
      })),
  );
  return [...fromGaps, ...fromRows];
}
